Major refactoring

domenkozar opened the pull request:
I've made a bunch of commit in last two days. I hope it's not too much of a hassle, but the general idea is to make howoldis more transparent about the whole process how we get a channel from a git commit.

Major improvements:
- replace `TagSoup` with `Haquery` (jQuery api in Haskell)
- use OverloadedStrings extension to simplify string types (a bit)
- enable responsive design for mobile users
- add three step explanation how channels are updated
- fetches each channel with HTTP HEAD to extract the link to the channel and thus also git commit
- adds a simple JSON view to fix #12
- add new fields to DiffChannel: times, channel link, commit hash, jobset name

A bit more information is in each commit :)

After this is online, I'll talk to Rob to host it at https://channel-status.nixos.org or similar (if that's fine with you).

![2016-10-18-204051_1001x884_scrot](https://cloud.githubusercontent.com/assets/126339/19491452/4ae4f972-9573-11e6-888d-8108bed87557.png)

cc @madjar

domenkozar committed:
We can lift Either with just a functor

domenkozar committed:
Rewrite tagsoup code to haquery

domenkozar committed:
Use overloadedstrings extension

domenkozar committed:
Add tests information inside DiffChannel

domenkozar committed:
Enable responsive design

domenkozar committed:
Cleanup Main.hs

domenkozar committed:
Add explanation how channel updates

domenkozar committed:
Request each channel link to extract more information

Each channel is requested using HTTP HEAD to extract
where it points.

From the url, git commit is extracted.

As a consequence, channel links are now pointing to cloudfront,
before they were a relative url.

Since requesting ~10 channels is slow, we do it in parallel.
I've done some measurements and once web server is warmed up
it usually took ~800ms.

domenkozar committed:
Display git commit on the frontend

domenkozar committed:
Add /api/channels JSON view

Also renames Channel to RawChannel and DiffChannel to Channel.

This is to avoid prefixing JSON API objects with d.

FRidh commented:
> After this is online, I'll talk to Rob to host it at https://channel-status.nixos.org or similar

and an alias at `howoldis.nixos.org` :-)

NeQuissimus commented:
Could we give the last commit a time stamp as well? It would make spotting Hydra issues much easier. The red blocks for 15.09 and earlier suggest some sort of badness, despite there simply not being a newer commit. (yes, 15.09 is eol...)

domenkozar commented:
@NeQuissimus Agreed. This would mean we'd have to use some sort of persistance since it will need 1s+ to collect all data. I'll do that once we host howoldis on nixos.org and can use some DB connection.

domenkozar commented:
@acowley if you can review this code I'd greatly appreciate, but I'd understand if I'm asking too much.

Will expand the two vague  `li`s

madjar commented:
@domenkozar That looks good to me. Since you spend more time than me on this code these days, I'm giving you the keys to the repo :) Feel free to merge this when you want!

domenkozar commented:
I'd like feedback on two things before I merge:
- general feedback on the code, for example if strings can be handled in a better way
- if text describing how channels are updated is clear enough

madjar commented:
I have not much to say about the strings. It's always annoying to jungle with Text and String. Maybe it could be made easier with [text-conversions](https://www.stackage.org/lts-7.4/package/text-conversions-0.3.0) and [mono-traversable](https://www.stackage.org/lts-7.4/package/mono-traversable-1.0.1), but I'm not sure at all.

For the explanation, I'm not sure I understand what an aggregate job is. Also, you seem to insist on the difference between unstable and tested, but I'm not sure I understand the how they differ. Finally, "jobset" might be a jargon word that needs explaining.

Instead of persistence, maybe we can have a simple in-memory cache, like [auto-update](https://hackage.haskell.org/package/auto-update-0.1.4/docs/Control-AutoUpdate.html) (which ensure the data collection is only run once every n seconds, and cache the result in-between).

copumpkin commented on Channels.hs:
```diff
@@ -1,59 +1,101 @@
+
+makeChannel :: UTCTime -> RawChannel -> IO Channel
+makeChannel current channel = do
+  e <- try $ W.head_ $ unpack $ "http://nixos.org/channels/" <> (rname channel)
```
Unnecessary parens on `(rname channel)`. Yes, I provide very profound feedback 😄

copumpkin commented on Channels.hs:
```diff
@@ -62,21 +104,19 @@ humanTimeDiff d
 
 -- | Takes time since last update to the channel and colors it based on it's age
-diffToLabel :: Either String NominalDiffTime -> Color
+diffToLabel :: Either Text NominalDiffTime -> Color
```
I know you didn't write this part, but "success", "warning", and "danger" don't really feel like colors to me, and it seems like it could be a simple sum/enum instead of `Text`

domenkozar replied:
So you mean something like `data Label = Danger | Warning | Success` and then a function that maps those to `Text`?

acowley commented:
The code is high quality. The only thing that jumped out at me was the use of `Text` as a representation of `Color` (as @copumpkin already pointed out) when an enumeration should probably be used as an intermediary between the logic and presentation.

The prose still needs some grammar smoothing, but I think the jargon consistency is the larger problem. You define "unstable" and "tested" **jobs**, then refer to them later as **jobsets**. Perhaps it is worth spelling things out further (and I may be wrong in my own apprehension here). 

Each channel has an associated `jobset` that is a collection of individual `job`s. A `job` roughly corresponds to building a particular package, which itself may depend on other packages built by other `job`s. Before a channel can update, its associated `jobset` must be finished building, though the building of some constituent `job`s may fail. As long as certain special `job`s -- `unstable` for `nixpkgs`, and `tested` for `nixos` -- build successfully, the channel can update. The role of these signal `job`s is to ensure that essential functionality is present on each channel update.

copumpkin commented on Main.hs:
```diff
@@ -3,33 +3,24 @@
-                                <|> Just (head chans)
-  where lookup n = find (\c -> dname c == n) chans
+    get "/api/channels" $ do
+      allChannels <- liftIO $ channels
```
Why all this `foo $ bar` instead of `foo bar`?

domenkozar replied:
HLint fully agress, pushed.

domenkozar committed:
HLint cleanup

domenkozar committed:
Color is now a type safe Label

domenkozar committed:
Note API link in the footer

domenkozar commented:
Thanks all for feedback!
- added `Label` type constructor
- cleaned up source with `HLint`
- added a link to the only API JSON we have

@madjar I'd like to move towards using a DB sooner or later to keep a history of channel updates. It will be useful to analyze them (even more so once we ship security updates properly).

I still have to address @acowley excellent comment about my rhetoric.

domenkozar committed:
README: mention how to install

domenkozar committed:
Wording based on @acowley feedback

domenkozar commented:
I've pushed a wording update based on @acowley feedback. If everyone can go through the three sections and comment if something is unclear I'd appreciate this:

![2016-10-20-150744_679x90_scrot](https://cloud.githubusercontent.com/assets/126339/19560979/07fa8bee-96d7-11e6-9d0c-f4b76d310cf4.png)
![2016-10-20-150756_953x365_scrot](https://cloud.githubusercontent.com/assets/126339/19560978/07f8b350-96d7-11e6-805a-e77e57221c04.png)
![2016-10-20-150806_939x274_scrot](https://cloud.githubusercontent.com/assets/126339/19560980/07fb3eea-96d7-11e6-8a13-5b8e4b7c676c.png)

PS: feel free to be a grammar-nazi

NeQuissimus commented:
I would change a few things. Strangely, they're all in section 3 :D

Changes are bold

3) 

A particular jobset evaluation needs to be complete **built,** i.e. no more queued jobs, ...
A particular jobset evaluation**'s** tested/unstable job needs ...

**The** nixos.org server ... poll for **the** newest jobset ...
Once triggered, release likes **such as** ISOs are copied**. For the** NixOS channel, the command-not-found index is generated...

... above needs to happen once **a** channel update is triggered

domenkozar committed:
Grammar

domenkozar commented:
@madjar will this autodeploy once merged?

FRidh commented:
I think we should get rid of those colors. The use of red, yellow and green typically imply that there is something right and wrong. However, red doesn't have to mean something is wrong. Instead I propose to use a single base color and use different shades, e.g., more bright when it is newer and less bright when it is older. You could use here say the logarithm of the amount of days to keep most of the bandwidth for the first couple of days. I think the shades of blue would fit in nicely.

madjar commented:
@domenkozar Yep, there is some magic setup on heroku, so it should autodeploy in about the time it takes to recompile.

domenkozar commented:
@FRidh I agree. However I'd like to fix this with https://github.com/madjar/howoldis/issues/16, then red/yellow will have an actually meaning with the information how long ago there was a newer commit, but no channel update.

domenkozar commented:
Merging this for now, as I think it improve the service quite a bit. If you have any feedback, feel free to leave it here!
